Source/ApparentHorizonFinder/ApparentHorizon.impl.hpp

```diff
@@ -0,0 +1,1848 @@
+    {
+        SNESConvergedReason reason = solver.getConvergedReason();
+
+        // result will be 0 if any of the PETSc ranks says 'reason <=0' (PETSc
+        // convergence error)
+        int tmp = reason; // convert just to make sure it's 'int'
+#ifdef CH_MPI
+        MPI_Allreduce(&tmp, &result, 1, MPI_INT, MPI_MIN, Chombo_MPI::comm);
+#else
+        result = tmp;
+#endif
+    }
+    else
+    {
+        int ALOT = 100; // smaller than all SNESConvergedReasons
+#ifdef CH_MPI
+        MPI_Allreduce(&ALOT, &result, 1, MPI_INT, MPI_MIN, Chombo_MPI::comm);
+#else
+        result = ALOT;
+#endif
+    }
+
+    bool converged = (result > 0);
+
+    if ((bool)converged)
+    {
+        ++m_converged;
+        m_num_failed_convergences = 0;
+    }
+    else
+    {
+        m_converged = 0; // reset
+        ++m_num_failed_convergences;
+    }
+
+    if (m_params.verbose > AHParams::NONE)
+    {
+        pout() << (m_converged ? "Solver converged. Horizon FOUND."
+                               : "Solver diverged. Horizon NOT found.")
+               << std::endl;
+
+        if (m_params.verbose > AHParams::MIN)
+            pout() << "SNESConvergedReason = " << result << std::endl;
+
+        if (result == SNES_DIVERGED_MAX_IT)
+            pout() << "(maximum iterations reached, try increasing "
+                      "AH_SNES_max_iterations)"
+                   << std::endl;
+    }
+
+    if (!m_has_been_found && m_converged)
+        m_has_been_found = true; // finally found :D
+}
+
+template <class SurfaceGeometry, class AHFunction>
+void ApparentHorizon<SurfaceGeometry, AHFunction>::restart(
+    bool solve_first_step)
+{
+    CH_TIME("ApparentHorizon::restart");
+
+    static double eps = 1.e-7;
+
+    // GUIDING PRINCIPLE : restart should not depend on m_params, as these might
+    // have changed! We should be able to figure everything out without them
+
+    const GRAMR &gramr = (static_cast<const GRAMR &>(
+        solver.m_interp.get_interpolator()->getAMR()));
+
+    int current_step = AMR::s_step;
+    int restart_step = gramr.get_restart_step();
+
+    bool solve_time_0_now = (!m_params.re_solve_at_restart &&
+                             current_step == 0 && restart_step < 0);
+
+    // solve if at t=0 and not a restart
+    if (solve_time_0_now)
+    {
+        // 'dt' doesn't matter for 1st time step, so set to 1
+        if (solve_first_step)
+            solve(1., 0., 0.);
+        return;
+    }
+
+    double current_time = gramr.getCurrentTime();
+    double coarse_dt = (current_step == 0 ? 0. : current_time / current_step);
+    double level_dt = coarse_dt * pow(2., -m_params.level_to_run);
+    double current_solve_dt = level_dt * m_params.solve_interval;
+
+    // READ STATS
+
+    // get centre from stats file
+    std::string file = m_params.stats_path + m_stats + ".dat";
+    std::vector<std::vector<double>> stats;
+    SmallDataIOReader stats_reader;
+    stats_reader.open(file);
```

Comment:
This fails in the case where the file does not exist - e.g. for the binary where the 3rd AH has not appeared yet. If I create an empty file called `stats_AH3.dat` then it work ok, as it triggers the if statement just below, but this should also happen for the case where no file exists - currently it just fails at this open statement.

Reply:
@KAClough by number of points are you referring to AH_num_points_u and AH_num_points_v?

Reply:
Yes right - actually I just changed the u points